cli/app/api.go

```diff
@@ -442,45 +443,520 @@ func PublishToMashery(user *ApiUser, appDir string, gatewayJSON string, host str
+	updated := false
 
-	if cleanedTfSwaggerDoc, err = json.Marshal(m["document"]); err != nil {
+	var apiId string
+	var apiName string
+	var endpoints [](interface{})
+	//f
+	api, err := user.Read("services", "name:"+mApi["name"].(string), "id,name,endpoints.id,endpoints.name,endpoints.inboundSslRequired,endpoints.outboundRequestTargetPath,endpoints.outboundTransportProtocol,endpoints.publicDomains,endpoints.requestAuthenticationType,endpoints.requestPathAlias,endpoints.requestProtocol,endpoints.supportedHttpMethods,endoints.systemDomains,endpoints.trafficManagerDomain", token)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error: Unable to fetch api\n\n")
 		panic(err)
 	}
 
-	if mock == false {
-		s, err := user.CreateAPI(string(cleanedTfSwaggerDoc), token)
+	var f [](interface{})
+	if err = json.Unmarshal([]byte(api), &f); err != nil {
+		panic(err)
+	}
+	if len(f) == 0 {
+		s, err := user.Create("services", "id,name,endpoints.id,endpoints.name,endpoints.inboundSslRequired,endpoints.outboundRequestTargetPath,endpoints.outboundTransportProtocol,endpoints.publicDomains,endpoints.requestAuthenticationType,endpoints.requestPathAlias,endpoints.requestProtocol,endpoints.supportedHttpMethods,endoints.systemDomains,endpoints.trafficManagerDomain", string(cleanedTfApiSwaggerDoc), token)
 		if err != nil {
 			fmt.Fprintf(os.Stderr, "Error: Unable to create the api %s\n\n", s)
 			fmt.Errorf("%v", err)
-			return err
+			panic(err)
 		}
+		apiId, apiName, endpoints = GetApiDetails(s)
 
-		fmt.Println("Successfully published to mashery!")
 	} else {
-		var prettyJSON bytes.Buffer
-		err := json.Indent(&prettyJSON, cleanedTfSwaggerDoc, "", "\t")
+		m := f[0].(map[string]interface{})
+		var m1 map[string]interface{}
+		json.Unmarshal(cleanedTfApiSwaggerDoc, &m1)
+		merged := merge(m, m1, 0)
+		var mergedDoc []byte
+		if mergedDoc, err = json.Marshal(merged); err != nil {
+			panic(err)
+		}
+		serviceId := merged["id"].(string)
+		s, err := user.Update("services/"+serviceId, "id,name,endpoints.id,endpoints.name,endpoints.inboundSslRequired,endpoints.outboundRequestTargetPath,endpoints.outboundTransportProtocol,endpoints.publicDomains,endpoints.requestAuthenticationType,endpoints.requestPathAlias,endpoints.requestProtocol,endpoints.supportedHttpMethods,endoints.systemDomains,endpoints.trafficManagerDomain", string(mergedDoc), token)
 		if err != nil {
-			return err
+			fmt.Fprintf(os.Stderr, "Error: Unable to update the api %s\n\n", s)
+			fmt.Errorf("%v", err)
+			panic(err)
 		}
+		apiId, apiName, endpoints = GetApiDetails(s)
 
-		fmt.Printf("%s", prettyJSON.Bytes())
-		fmt.Println("\nMocked! Did not attempt to publish.\n")
+		updated = true
 	}
 
-	return nil
+	return apiId, apiName, endpoints, updated
+}
+
+func merge(dst, src map[string]interface{}, depth int) map[string]interface{} {
+	for key, srcVal := range src {
+		if dstVal, ok := dst[key]; ok {
+			if reflect.ValueOf(dstVal).Kind() == reflect.Map {
```

Comment:
Would a type switch work here instead of reflection: https://tour.golang.org/methods/16